### PR TITLE
Add staged contract migration

### PR DESCRIPTION
**Pull request by SupunS**
Closes onflow/cadence#3038

### CODE REVIEW

File: cmd/util/ledger/migrations/staged_contracts_migration.go [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+		oldProgram,
+		newProgram,
+	)
+
+	return validator.Validate()
+}
+
+func GetStagedContracts() []StagedContracts {
+	// TODO:
+	return nil
```
Comment (SupunS) [inline cmd/util/ledger/migrations/staged_contracts_migration.go:206]: TBD: To be constructed from the list/csv obtained from the staged contract deployment. See: https://discord.com/channels/613813861610684416/1108479699732152503/1200140480269254696

File: go.mod [hunk 2]
```diff
@@ -51,7 +51,7 @@ require (
 	github.com/multiformats/go-multiaddr-dns v0.3.1
 	github.com/multiformats/go-multihash v0.2.3
 	github.com/onflow/atree v0.6.1-0.20230711151834-86040b30171f
-	github.com/onflow/cadence v1.0.0-M3
+	github.com/onflow/cadence v1.0.0-preview.2.0.20240125192801-40b4d6928fcf
```
Comment (SupunS) [inline go.mod:54]: https://github.com/onflow/cadence/pull/3043 got merged after the `v1.0.0-M3` tag. So bumped the cadence version to the commit 40b4d6928fcfcd334a62dc934549ff32a5b9f0f7

Comment (turbolent) [reply go.mod:54]: In an unrelated PR I'd also want https://github.com/onflow/cadence/pull/3042. 

Let's just create a new tag and update to that one

**Review by turbolent — APPROVED** [review]
Great work!

File: cmd/util/cmd/execution-state-extract/execution_state_extract.go [hunk 3]
```diff
@@ -98,6 +98,9 @@ func extractExecutionState(
 				log,
 				nWorker,
 				[]migrators.AccountBasedMigration{
+					// Contracts must be migrated first
+					migrators.NewStagedContractsMigration(migrators.GetStagedContracts),
```
Comment (SupunS) [inline cmd/util/cmd/execution-state-extract/execution_state_extract.go:102]: I added this for the moment. But we would need to run the staged-contracts migration for all accounts first, before running the value migrations for any account. Not sure how to do it in a single binary. Can add it in a separate PR.

Comment (SupunS) [reply cmd/util/cmd/execution-state-extract/execution_state_extract.go:102]: cc @janezpodhostnik

File: cmd/util/ledger/migrations/staged_contracts_migration.go (hunk 1 above)
Comment (turbolent) [reply cmd/util/ledger/migrations/staged_contracts_migration.go:205]: Can you please create an issue for that remainig work? We can definitely get this PR in before, as it depends on Gio's work on the staged contract update contract. Maybe define an "interface"/format – maybe we can read in JSON-CDC?

Comment (SupunS) [reply cmd/util/ledger/migrations/staged_contracts_migration.go:205]: https://github.com/onflow/cadence/issues/3050

File: cmd/util/cmd/execution-state-extract/execution_state_extract.go [hunk 4]
```diff
@@ -94,6 +94,15 @@ func extractExecutionState(
...
 
 		migrations = []ledger.Migration{
+			// Contracts must be migrated first
+			migrators.CreateAccountBasedMigration(
+				log,
+				nWorker,
+				[]migrators.AccountBasedMigration{
+					migrators.NewStagedContractsMigration(migrators.GetStagedContracts),
+				},
+			),
```
Comment (SupunS) [inline cmd/util/cmd/execution-state-extract/execution_state_extract.go:104]: Not sure if this is the proper way to do this: We need to run contract migration for all accounts, before running other migrations.

**Review by turbolent — APPROVED** [review]

**Comment by turbolent** [comment]
`TestCadenceValuesMigration` is failing, not quite sure why

**Comment by turbolent** [comment]
Sorry, that was my bad, I had changed the format of the error message!

**Comment by SupunS** [comment]
ah no, that's totally fine! It's a bit fragile/not really a good practice to assert on the error messages anyway. But don't really have a better alternative in this case really. In that test I want to make sure the 'correct' error is on the correct account, path, and the key (which is pretty much the entire error message)

**Pull request merged**